### PR TITLE
make sure the msrv for `const_raw_ptr_deref` is met when linting [`missing_const_for_fn`]

### PR DESCRIPTION
**Pull request by J-ZhengLi**
fixes: #8864 

---

changelog: make sure the msrv for `const_ptr_deref` is met when linting [`missing_const_for_fn`]

### CODE REVIEW

File: clippy_lints/src/missing_const_for_fn.rs [hunk 1]
```diff
@@ -149,6 +151,11 @@ impl<'tcx> LateLintPass<'tcx> for MissingConstForFn {
             return;
         }
 
+        // Skip if there is any pointer deref ops in the body, and the msrv of that feature does not meet.
+        if !self.msrv.meets(msrvs::CONST_RAW_PTR_DEREF) && has_raw_ptr_deref_in_body(cx, body) {
```
Comment (y21) [inline clippy_lints/src/missing_const_for_fn.rs:155]: Should this logic be in `is_min_const_fn` instead? Other lints also use that to check if something can be `const`:
```rs
thread_local! {
    static V: u8 = unsafe { *(&1u8 as *const u8) };
}
```
^ This probably also can't be made `const` for old MSRVs?

Comment (J-ZhengLi) [reply clippy_lints/src/missing_const_for_fn.rs:155]: well, I'm not familiar with those mir stuffs, so I don't know where should I put the equivalent logic in that function.

and that thread_local lint was added in `1.77` so I think it should be fine naturally~

Comment (y21) [reply clippy_lints/src/missing_const_for_fn.rs:155]: probably in this match https://github.com/rust-lang/rust-clippy/blob/master/clippy_utils/src/qualify_min_const_fn.rs#L270

`ProjectionElem::Deref` if `base.ty().is_unsafe_ptr()` should be it I think.

**Comment by J-ZhengLi** [comment]
Finally got some time to work on this. @rustbot ready

File: clippy_utils/src/qualify_min_const_fn.rs [hunk 2]
```diff
@@ -245,33 +253,33 @@ fn check_operand<'tcx>(tcx: TyCtxt<'tcx>, operand: &Operand<'tcx>, span: Span, b
...
-                    // No union field accesses in `const fn`
-                    if def.is_union() {
-                        return Err((span, "accessing union fields is unstable".into()));
-                    }
+                if base.ty(body, tcx).ty.is_union() && !msrv.meets(msrvs::CONST_FN_UNION) {
+                    return Err((span, "accessing union fields is unstable".into()));
+                }
+            },
+            ProjectionElem::Deref => {
+                if base.ty(body, tcx).ty.is_unsafe_ptr() && !msrv.meets(msrvs::CONST_RAW_PTR_DEREF) {
```
Comment (y21) [inline clippy_utils/src/qualify_min_const_fn.rs:275]: Dereferencing `*mut T` is still unstable today (https://github.com/rust-lang/rust/issues/57349) so I think this should only allow `RawPtr(Mutability::Not)` specifically and always error for `RawPtr(Mutability::Mut)`
(sorry if I suggested `is_unsafe_ptr` in the first place 😅)

**Comment by y21** [comment]
Since dswij seems to be a little busy and has a bunch of other PRs assigned and I already started reviewing this, I'm gonna steal this one to move it forwards. Hope you dont mind :D 
r? @y21

File: clippy_utils/src/qualify_min_const_fn.rs (hunk 2 above)
Comment (J-ZhengLi) [reply clippy_utils/src/qualify_min_const_fn.rs:275]: Oh, right, I forgot about the mut pointers as well 😄

**Comment by dswij** [comment]
Yeah, feel free :)

**Review by y21 — APPROVED** [review]
LGTM

**Comment by y21** [comment]
Thanks! @bors r+

**Pull request merged**